reinstate phpcs php linting

haszari opened the pull request:
Prompted by #1393
Fixes #1356

Update: woah cool there is an issue for this! #1356 🎩 @nielslange 

The original goal of this PR was to update the WordPress coding standards PHPcs ruleset, and anything necessary related to it.

I discovered that our phpcs setup wasn't in working order, so have upgraded and replaced various packages. 

~~Work in progress - need to ensure we have enough rules in place that match the existing code. Currently have lots of violations.~~ All violations now fixed and pre-commit should catch issues in new code.

Summary of what's included in this PR:

- Updated and set versions for various phpcs related packages. Key outcome here is getting on the latest WordPress standards. fac37e55bfb523c07e37299edf46946839290d5f
- Added/tweaked commands in composer.json and package.json for visibility and to encourage use/familiarity.
- Set min PHP version of sniffer config to match plugin min PHP version.
- Mega commit fixing all formatting issues automatically using `phpcbf`. 7c90d2311b0433ee9cb0f9d896704c64db3ae0e5
- Manual fixing of a few more interesting violations (~10 issues). One commit per file for visibility. Note this included:
  - Adding/fixing version param in all uses of `wp_enqueue_style`. This means that various third-party assets will be cached less aggressively (now tied to `$storefront_version`).
  - Missing escaping of comment dates in `storefront_comment()` (🧤)
  - Use `gmdate()` instead of ambiguous `date()`.
- Fixed one rule ignore comment to latest format. (Still ignoring, didn't investigate why)

##### We still have some warnings throughout code
> Also notice that when I run that command I get several warnings. I assume that's expected? – @Aljullu 

I've fixed all errors but have not touched any warnings. Warnings should be _carefully_ handled as follow up - linter warnings can be very subtle/tricky/risky  🐲 to fix, so I'm not rushing to fix them all in this PR!

**Update** I have bulk-fixed an older style of ignore comment to reduce the amount of open warnings. These comments opt-out of XSS checks, and we have a lot of them. c2b82fac62038925df1db0bf0fb6f7ed555808c9

- Old comment `// WPCS: XSS ok.`
- Current comment, using `phpcs:ignore` `// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped`


##### Travis / CI is _not_ included (yet)
Note there's nothing in this PR to trigger linter check as a CI step (e.g. on PRs). We should do that next! #1395 

### How to test the changes
#### Test linter tooling
1. Get on this branch, `composer install`, `npm install`.
   - Note: `composer install` is what sets up the `pre-commit` hook (TIL).
2. Run linter `npm run lint:php`. Should pass with flying colours (* and plenty of warnings 😿).
3. Write some [bad PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/).
4. Run linter again - should tell you what's wrong.
5. Try to commit the bad changes - commit hook should catch it. 
5. Write some PHP with auto-fixable changes (e.g. spaces/tabs, dangling commas, long lines, multi-line func calls).
6. Run auto fix to `npm run lint:php:fix`. Should fix the errors so you can commit.

#### Broad smoke testing of a cross-section of Storefront functionality
As this is a big change, any testing of Storefront features would be appreciated. 

cc @woocommerce/rubik I've tagged you all for review – first in first served. I don't expect everyone to review – grateful for any reviews or testing.

If you could run this PR (or master after merge) in your dev environment and keep an eye out for issues I would really appreciate it 🙌 🙇 

#### Affected flows
- Everything (in theory). 
- Comment dates now escaped. 

### Changelog

> Dev – Reinstate phpcs PHP linting. #1463

haszari committed:
update squizlabs/php_codesniffer and WP coding standards to latest

haszari committed:
revamp phpcs sniffer rulesets:
- update to php-compatibility (wimg is defunct)
- add version for woocommerce-sniffs

haszari committed:
streamline phpcs config (get it working at all):
- update min WP version to 5.3
- remove references to rules we don't have installed
  - WordPress.VIP
  - WordPress.XSS

haszari committed:
fix 2x phpcs errors (automatically via phpcbf)

haszari committed:
update phps ignore unescaped output comment to current format

haszari committed:
add npm scripts for php linter and linter autofixer

haszari committed:
auto-fix PHP linter errors using phpcbf :

Before phpcbf:

[x] PEAR      Functions           Function call signature multiple arguments          127
[x] Generic   White space         Disallow space indent spaces used                   10
[ ] WordPres  WP                  Enqueued resource parameters missing version        6
[ ] PHPCompa  Language construct  New language constructs t_ns_separator found        4
[ ] WordPres  Security            Escape output output not escaped                    4
[ ] PHPCompa  Parameter values    New HTMLEntities encoding default not set           2
[ ] WordPres  Date time           Restricted functions date_date                      2
[x] Generic   Files               End file newline not found                          1
[x] PEAR      Functions           Function call signature close bracket line          1
[x] PEAR      Functions           Function call signature content after open bracket  1
[x] Squiz     White space         Superfluous whitespace end line                     1
[x] WordPres  Arrays              Comma after array item no comma                     1

phpcbf fixed all [x] violations, all whitespace/formatting

After phpcbf:

WordPres  WP                  Enqueued resource parameters missing version      6
PHPCompa  Language construct  New language constructs t_ns_separator found      4
WordPres  Security            Escape output output not escaped                  4
PHPCompa  Parameter values    New HTMLEntities encoding default not set         2
WordPres  Date time           Restricted functions date_date                    2

Note - this commit does not include auto-fixed files with other
violations. These will follow in separate commit (after fixing!)

haszari committed:
fix phpcs violations:
- numerous formatting issues fixed automatically
- manually fix missing version param in calls to wp_enqueue_style

haszari committed:
fix phpcs violations:
- numerous formatting issues fixed automatically
- fix missing deps param in call to wp_enqueue_style

haszari committed:
update phpcs test php version to match min requirement (5.6)

haszari committed:
fix phpcs violations including some missing escaping:
- numerous formatting issues fixed automatically
- prefer gmdate() over potentially ambiguous date()
- escape output (a real issue!) of comment dates

haszari committed:
fix violations (all automated formatting fixes)

Aljullu commented on phpcs.xml:
```diff
@@ -21,27 +21,6 @@
 		<exclude name="PHPCompatibility.PHP.NewKeywords.t_namespaceFound" />
 	</rule>
 
-	<rule ref="WordPress">
```
I'm missing why this rule was removed. Is it not adding anything? I tried re-enabling it locally to test it and doesn't seem to make any difference, but wanted to ask to be sure.

haszari replied:
That was unintentional … I'll put it back. I was hacking on this xml file heavily trying to confirm the right rulesets were applying, thanks for catching.

Aljullu reviewed (approved):
Nice work with this! Really great to see projects getting into a shared set of linting rules and code styles. I just left a comment below about a rule that was removed, but everything else looking and testing great. :ok_hand: 

@haszari FYI I updated the testing instructions in the PR description. They were pointing to `npm run php:lint` as the command to run, but it should be `npm run lint:php`.

Also notice that when I run that command I get several warnings. I assume that's expected?

haszari committed:
reinstate WordPress rule/standard in phpcs.xml (minimise PR changes)

haszari committed:
exclude build (pre-zip) ./storefront & tighten excludes for dep folders

haszari committed:
bulk-update Security.EscapeOutput.OutputNotEscaped ignore comment:
- the previous comment format is no longer supported
- bulk replacing these to reduce phpcs warning overhead

haszari commented:
I've done some broad smoke testing and didn't see any issues - so merging. Please all keep an eye out for any impact of this change (🤞).
